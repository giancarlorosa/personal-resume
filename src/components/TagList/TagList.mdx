import { Meta, Canvas, Controls } from '@storybook/blocks';
import * as TagListStories from './TagList.stories.jsx';

<Meta of={TagListStories} />

# Tag List Component

The **Tag List** component was developed to list an undetermined number of items inside its container. These internal elements are called **Tags** and receive **children** parameter in a **string** format. The component could be built using a parameter as an array, but to set the component more readable, it was decided to build a compound component. The implementation is very simple, where a **TagList** container is set, and for any new internal item, add a component called **Tag**.

## Implementation

This component has a very simple implementation. Just inform a **Tag** subcomponent as children.

<Canvas of={TagListStories.Implementation} />

<Controls of={TagListStories.Implementation} />

## Multiple Lines

Here it's possible to test the multiple lines inside the component to check its margin.

<Canvas of={TagListStories.MultipleLines} />
