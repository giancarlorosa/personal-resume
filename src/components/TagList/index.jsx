import PropTypes from 'prop-types';
import classNames from 'classnames';

function TagList({ children }) {
  return <ul>{children}</ul>;
}

function Tag({ children }) {
  const tagClasses = classNames(
    'bg-deep-sky-blue',
    'text-white',
    'inline-block',
    'py-1.5',
    'px-2.5',
    'my-1',
    'mx-0.5',
    'rounded'
  );
  return <li className={tagClasses}>{children}</li>;
}

// TODO: find a way to validate compound components.
TagList.propTypes = { children: PropTypes.node.isRequired };
Tag.propTypes = { children: PropTypes.string.isRequired };

export default TagList;
export { Tag };
