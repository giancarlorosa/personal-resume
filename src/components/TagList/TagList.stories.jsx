import TagList, { Tag } from '.';

// TODO: find a way to document a compound component.
export default {
  title: 'Components/Tag List',
  component: TagList,
};

export const Implementation = {
  render: () => (
    <TagList>
      <Tag>Marketing</Tag>
      <Tag>Print</Tag>
      <Tag>Digital</Tag>
      <Tag>Design</Tag>
      <Tag>Social</Tag>
      <Tag>Media</Tag>
      <Tag>Time</Tag>
      <Tag>Management</Tag>
      <Tag>Communication</Tag>
      <Tag>Problem-Solving</Tag>
      <Tag>Social</Tag>
      <Tag>Networking</Tag>
      <Tag>Flexibility</Tag>
    </TagList>
  ),
};

export const MultipleLines = {
  render: () => (
    <div className="max-w-lg">
      <TagList>
        <Tag>Marketing</Tag>
        <Tag>Print</Tag>
        <Tag>Digital</Tag>
        <Tag>Design</Tag>
        <Tag>Social</Tag>
        <Tag>Media</Tag>
        <Tag>Time</Tag>
        <Tag>Management</Tag>
        <Tag>Communication</Tag>
        <Tag>Problem-Solving</Tag>
        <Tag>Social</Tag>
        <Tag>Networking</Tag>
        <Tag>Flexibility</Tag>
      </TagList>
    </div>
  ),
};
