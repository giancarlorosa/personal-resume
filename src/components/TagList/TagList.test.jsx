/* eslint-disable no-console */
import { render } from '@testing-library/react';
import TagList, { Tag } from '.';

describe('Section Title', () => {
  beforeAll(() => jest.spyOn(console, 'error').mockImplementation(() => {}));
  afterAll(() => console.error.mockRestore());
  afterEach(() => console.error.mockClear());

  it('should render the component', () => {
    const { getByText } = render(
      <TagList>
        <Tag>Tag Item 1</Tag>
        <Tag>Tag Item 2</Tag>
        <Tag>Tag Item 3</Tag>
      </TagList>
    );
    const tagItemOne = getByText('Tag Item 1');
    const tagItemTwo = getByText('Tag Item 2');
    const tagItemThree = getByText('Tag Item 3');

    expect(tagItemOne).toBeTruthy();
    expect(tagItemTwo).toBeTruthy();
    expect(tagItemThree).toBeTruthy();
  });

  it('should contain ul and li tag', () => {
    const { getByRole, getAllByRole } = render(
      <TagList>
        <Tag>Tag Item 1</Tag>
        <Tag>Tag Item 2</Tag>
        <Tag>Tag Item 3</Tag>
      </TagList>
    );

    const tagList = getByRole('list');
    const tagItems = getAllByRole('listitem');

    expect(tagList).toBeTruthy();
    expect(tagItems).toBeTruthy();
    expect(tagList.tagName).toBe('UL');
    expect(tagItems[0].tagName).toBe('LI');
    expect(tagItems.length).toBe(3);
  });

  it.each([
    {
      errorLabel: 'TagList children prop is not set',
      expectText: 'The prop `children` is marked as required in `TagList`',
      tags: [],
    },
    {
      errorLabel: 'Tag children prop is not set',
      expectText: 'The prop `children` is marked as required in `Tag`',
      tags: [undefined],
    },
    {
      errorLabel: 'Tag children prop is not a string',
      expectText: 'Invalid prop `children` of type `object` supplied to `Tag`',
      tags: [<div>Tag Item</div>],
    },
  ])('should throw an error when $errorLabel', ({ expectText, tags }) => {
    if (tags.length) {
      render(
        <TagList>
          {tags.length > 0 && tags.map((tag) => <Tag>{tag}</Tag>)}
        </TagList>
      );
    } else {
      render(<TagList />);
    }

    expect(console.error).toHaveBeenCalled();
    expect(console.error.mock.calls[0][2]).toContain(expectText);
  });

  it('should verify changes on Tag component styles', () => {
    const { getByText } = render(
      <TagList>
        <Tag>Tag Item 1</Tag>
      </TagList>
    );

    const tagItem = getByText('Tag Item 1');
    expect(tagItem.className).toMatchSnapshot();
  });
});
