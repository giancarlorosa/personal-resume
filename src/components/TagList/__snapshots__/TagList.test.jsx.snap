// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section Title should verify changes on Tag component styles 1`] = `"bg-deep-sky-blue text-white inline-block py-1.5 px-2.5 my-1 mx-0.5 rounded"`;
